### ISSUE TITLE
pipeline.pl should print help 

### ISSUE THREAD

**Issue by lewismc** [joshua-decoder/joshua#238]
It would be very handy if pipeline.pl prints all command line options if invoked without parameters or with --help.
This issue will address that.

**Comment by mjpost** [comment]
Agreed. There isn't an easy way to do this under the current setup, because of the way the pipeline handles arguments (barebones home-brewed).

This is tied to a more complex issue, which is that the pipeline needs to be rewritten entirely, more in the spirit of Moses' EMS.

**Comment by lewismc** [comment]
... bugger.
I had a look at [Pod::Usage](http://perldoc.perl.org/Pod/Usage.html) which seems like it would do the trick.
pipeline.pl is pretty meaty... I am not sure that I want to embark on a rewrite based on the fact that I don't know the code too well (and the fact that it's in pl).
What's your thoughts here @mjpost ?

**Comment by mjpost** [comment]
Pipeline started out as a small Perl script and, as they tend to do, grew. Beside being a huge Perl script (itself a big problem), it has a number of other issues. It would be good to start with a graph-based approach to the needs of the MT pipeline, start at the end, and work backwards. This is the way tools like Make or Moses' EMS work. Pipeline starts at the beginning each time, and tries to be smart about what steps it can skip.

This might be a situation where it's better to just work with what you've got. But I'd be happy to take a tool like EMS, get permission to Apache-2 it, and then roll it into Joshua. I'm reluctant to put too much work into Pipeline.

(I also have another option for build tools that I am looking into)

**Comment by lewismc** [comment]
@mjpost I understand, this sounds like a significant undertaking.
It looks like simply printing a help message for pipeline.pl is not even a trivial undertaking.
Are you referring to this [EMS](http://www.statmt.org/moses/?n=FactoredTraining.EMS)?

**Comment by mjpost** [comment]
Yep.
